<h2 id="running-alevin">Running Alevin:</h2>

<p>Alevin has multiple features and modes which based on the requirements and use-case can toggled on and off.</p>

<h3 id="default-mode">Default Mode</h3>
<p>Once all the above resource are available alevin can be run using the following command:</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>./bin/salmon alevin -lISR -1 fastqs/pbmc4k_S1_L001_R1_001.fastq.gz -2 fastqs/pbmc4k_S1_L001_R2_001.fastq.gz --chromium -i index -p 8 -o alevin_output --tgMap txp2gene.tsv
</code></pre>
</div>

<h3 id="external-whitelist">External Whitelist</h3>
<p>Many times (specifically in case of 10x data), the whitelist is already available and based on the use-case, one don’t wan’t alevin to perform whitelisting of it’s own. This mode can be activated by giving <code class="highlighter-rouge">--whitelist</code> flag to alevin and providing a list of known whitelisted CB (Cellular Barcodes). Given, this flag alevin will sequence correct CB towards the given CB and generated the gene-count matrix accordingly.</p>

<h3 id="10x-v3-data">10x v3 Data</h3>
<p>The data format is almost similar to v2, the only chnage we have to do is to use <code class="highlighter-rouge">--chromiumV3</code> instead of <code class="highlighter-rouge">--chromium</code>.</p>

<h3 id="drop-seq-data">Drop-seq Data</h3>
<p>Alevin can be easily ported to quantify the data from <em>drop-seq</em> protocol as well. Since the data format is almost similar to <code class="highlighter-rouge">10x</code>, the only chnage we have to do is use <code class="highlighter-rouge">--dropseq</code> instead of <code class="highlighter-rouge">--chromium</code>.</p>

<h3 id="10x-v1-data">10x v1 Data</h3>
<p>Alevin is designed to primarily work with the file-format having CB-UMI in file while the corresponding read-sequence in the other. However, the 10x’s <code class="highlighter-rouge">v1</code> chemistry does not follow the same convention and primarily have UMI and read-sequence in the same file, breaking the parsing format for alevin. To support the working of alevin we have written a wrapper script which takes in the <code class="highlighter-rouge">v1</code> chemistry data and feed it to alevin in the required format. Since the wrapper script is not optimized for performance one can observe a time-hit compared to analysis done on <code class="highlighter-rouge">v2</code> chemistry data.</p>

<p>To run alevin in <code class="highlighter-rouge">v1</code> mode the following three changes are required:</p>

<ol>
  <li>Using the flag <code class="highlighter-rouge">--gemcode</code> instead of <code class="highlighter-rouge">--chromium</code>.</li>
  <li>NO <code class="highlighter-rouge">-1</code>, <code class="highlighter-rouge">-2</code>; flags should be given instead <code class="highlighter-rouge">-b</code> flag which specifies the path to the parent directory containing the reads (*I1* and *RA*).</li>
  <li>The wrapper script has to be downloaded and compile using the following command:</li>
</ol>

<div class="language-python highlighter-rouge"><pre class="highlight"><code><span class="n">git</span> <span class="n">clone</span> <span class="n">git</span><span class="nd">@github.com</span><span class="p">:</span><span class="n">COMBINE</span><span class="o">-</span><span class="n">lab</span><span class="o">/</span><span class="n">salmon</span><span class="o">.</span><span class="n">git</span>
<span class="n">cd</span> <span class="n">salmon</span><span class="o">/</span><span class="n">scripts</span><span class="o">/</span><span class="n">v1_10x</span><span class="p">;</span>
<span class="n">g</span><span class="o">++</span> <span class="o">-</span><span class="n">std</span><span class="o">=</span><span class="n">c</span><span class="o">++</span><span class="mi">11</span> <span class="o">-</span><span class="n">O3</span> <span class="o">-</span><span class="n">I</span> <span class="o">../../</span><span class="n">include</span> <span class="o">-</span><span class="n">o</span> <span class="n">wrapper</span> <span class="n">wrapper</span><span class="o">.</span><span class="n">cpp</span> <span class="o">-</span><span class="n">lz</span>
</code></pre>
</div>

<p>Once compiled, the above command will generate a binary with the name <code class="highlighter-rouge">wrapper</code> and the path to this binary should be updated in the <code class="highlighter-rouge">run.sh</code> file. The command to quantify the <code class="highlighter-rouge">v1</code> data would be as follows (from inside the <code class="highlighter-rouge">v1_10x</code> folder):</p>

<div class="language-bash highlighter-rouge"><pre class="highlight"><code>./run.sh ./salmon alevin -lISR -b pbmc3k_fastqs/ --gemcode -i index -p 8 -o alevin_output --tgMap txp2gene.tsv
</code></pre>
</div>

<h3 id="ipython-notebook">Ipython Notebook</h3>
<p>Prefer to read ipython notebook ?
Check out the gist <a href="https://gist.github.com/k3yavi/c501705ed2d29b12b0d10cf78b3ed001">here</a>.</p>
